driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2282,6 +2266,255 @@ public void run() {
+                        case UP:
+                            Host upHost = metadata.getHost(address);
+                            if (upHost == null) {
+                                upHost = metadata.add(address);
+                                // If upHost is still null, it means we didn't know about it the line before but
+                                // got beaten at adding it to the metadata by another thread. In that case, it's
+                                // fine to let the other thread win and ignore the notification here
+                                if (upHost == null)
+                                    continue;
+                                futures.add(schedule(hostAdded(upHost)));
+                            } else {
+                                futures.add(schedule(hostUp(upHost)));
+                            }
+                            break;
+                        case ADDED:
+                            Host newHost = metadata.add(address);
+                            if (newHost != null)
+                                futures.add(schedule(hostAdded(newHost)));
+                            break;
+                        case DOWN:
+                            // Note that there is a slight risk we can receive the event late and thus
+                            // mark the host down even though we already had reconnected successfully.
+                            // But it is unlikely, and don't have too much consequence since we'll try reconnecting
+                            // right away, so we favor the detection to make the Host.isUp method more reliable.
+                            Host downHost = metadata.getHost(address);
+                            if (downHost != null)
+                                futures.add(execute(hostDown(downHost)));
+                            break;
+                        case REMOVED:
+                            Host removedHost = metadata.getHost(address);
+                            if (removedHost != null)
+                                futures.add(execute(hostRemoved(removedHost)));
+                            break;
+                    }
+                }
+                return Futures.allAsList(futures);
+            }
+
+            private ListenableFuture<?> execute(ExceptionCatchingRunnable task) {
+                return executor.submit(task);
+            }
+
+            private ListenableFuture<?> schedule(final ExceptionCatchingRunnable task) {
+                // Cassandra tends to send notifications for new/up nodes a bit early (it is triggered once
+                // gossip is up, but that is before the client-side server is up), so we add a delay
+                // (otherwise the connection will likely fail and have to be retry which is wasteful). This
+                // probably should be fixed C* side, after which we'll be able to remove this.
+                final SettableFuture<?> future = SettableFuture.create();
```

Comment:
Just noticed that CASSANDRA-8236 is marked fixed for version 2.2.0-beta1. Shouldn't we then only apply a delay if version < 2.2.0?